app/src/frontend/building/data-container.tsx

```diff
@@ -14,39 +41,55 @@ import InfoBox from '../components/info-box';
  *
  * @param WrappedComponent
  */
-const withCopyEdit = (WrappedComponent) => {
-    return class extends React.Component<any, any> { // TODO: add proper types
+const withCopyEdit = (WrappedComponent: React.ComponentType<CategoryViewProps>) => {
+    return class DataContainer extends React.Component<DataContainerProps, DataContainerState> {
         static propTypes = { // TODO: generate propTypes from TS
             title: PropTypes.string,
             slug: PropTypes.string,
             intro: PropTypes.string,
             help: PropTypes.string,
             inactive: PropTypes.bool,
-            building_id: PropTypes.number,
             children: PropTypes.node
         };
 
         constructor(props) {
             super(props);
 
             this.state = {
-                error: this.props.error || undefined,
-                like: this.props.like || undefined,
+                error: undefined,
                 copying: false,
                 keys_to_copy: {},
-                building: this.props.building
+                buildingEdits: {},
+                currentBuildingId: undefined,
+                currentBuildingRevisionId: undefined
             };
 
             this.handleChange = this.handleChange.bind(this);
-            this.handleCheck = this.handleCheck.bind(this);
+            this.handleReset = this.handleReset.bind(this);
             this.handleLike = this.handleLike.bind(this);
             this.handleSubmit = this.handleSubmit.bind(this);
-            this.handleUpdate = this.handleUpdate.bind(this);
 
             this.toggleCopying = this.toggleCopying.bind(this);
             this.toggleCopyAttribute = this.toggleCopyAttribute.bind(this);
         }
 
+        static getDerivedStateFromProps(props, state) {
+            const newBuildingId = props.building == undefined ? undefined : props.building.building_id;
+            const newBuildingRevisionId = props.building == undefined ? undefined : props.building.revision_id;
+            if(newBuildingId !== state.currentBuildingId || newBuildingRevisionId > state.currentBuildingRevisionId) {
+                return {
+                    error: undefined,
+                    copying: false,
+                    keys_to_copy: {},
+                    buildingEdits: {},
+                    currentBuildingId: newBuildingId,
+                    currentBuildingRevisionId: newBuildingRevisionId
+                };
+            }
+
+            return null;
+        }
+
```

Comment:
The static method `getDerivedStateFromProps` is the improved mechanism for making changes to state based on changes to props, and is safer than the deprecated methods like `componentWillReceiveProps`
The state reflects the currently selected building and its revision. If a new building gets selected, or the current building is updated, the edit- and copy-related state is cleared